src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -669,15 +163,662 @@ void IUtilities::updateImageViaStagingBuffer(
+        stagingBufferOffset += size;
+        stagingBufferOffset = core::alignUp(stagingBufferOffset, bufferOffsetAlignment);
+        const auto consumedMemory = stagingBufferOffset - initialOffset;
+        if(consumedMemory <= availableMemory)
+        {
+            availableMemory -= consumedMemory;
+            return true;
+        }
+        else
+        {
+            assert(false);
+            return false;
+        }
+    };
+
+    if(!addToCurrentUploadBufferOffset(0u)) // in order to fix initial alignment to bufferOffsetAlignment
+    {
+        return false;
+    }
+
+    // ! Current Region that may break down into smaller regions (the first smaller region is nextRegionToCopy)
+    const asset::IImage::SBufferCopy & mainRegion = regions[currentRegion];
+        
+    asset::TexelBlockInfo srcImageTexelBlockInfo(srcImageFormat);
+    asset::TexelBlockInfo dstImageTexelBlockInfo(dstImageFormat);
+        
+    const core::vector4du32_SIMD srcBufferByteStrides = srcImageTexelBlockInfo.convert3DBlockStridesTo1DByteStrides(mainRegion.getBlockStrides(srcImageTexelBlockInfo));
+
+    // ! We only need subresourceSize for validations and assertions about minImageTransferGranularity because granularity requirements can be ignored if region fits against the right corner of the subresource (described in more detail below)
+    const auto subresourceSize = dstImage->getMipSize(mainRegion.imageSubresource.mipLevel);
+    const auto subresourceSizeInBlocks = dstImageTexelBlockInfo.convertTexelsToBlocks(subresourceSize);
+        
+    // regionBlockStrides = <BufferRowLengthInBlocks, BufferImageHeightInBlocks, ImageDepthInBlocks>
+    const auto regionBlockStrides = mainRegion.getBlockStrides(dstImageTexelBlockInfo);
+    // regionBlockStridesInBytes = <BlockByteSize,
+    //                              BlockBytesSize * BufferRowLengthInBlocks,
+    //                              BlockBytesSize * BufferRowLengthInBlocks * BufferImageHeightInBlocks,
+    //                              BlockBytesSize * BufferRowLengthInBlocks * BufferImageHeightInBlocks * ImageDepthInBlocks>
+    const core::vector4du32_SIMD regionBlockStridesInBytes = dstImageTexelBlockInfo.convert3DBlockStridesTo1DByteStrides(regionBlockStrides);
+    auto texelBlockDim = dstImageTexelBlockInfo.getDimension();
+
+    // ! Don't confuse imageExtent with subresourceSize, imageExtent is the extent of the main region to copy and the subresourceSize is the actual size of dstImage 
+    const auto imageExtent = core::vector3du32_SIMD(mainRegion.imageExtent.width, mainRegion.imageExtent.height, mainRegion.imageExtent.depth);
+    const auto imageOffset = core::vector3du32_SIMD(mainRegion.imageOffset.x, mainRegion.imageOffset.y, mainRegion.imageOffset.z);
+    const auto imageOffsetInBlocks = dstImageTexelBlockInfo.convertTexelsToBlocks(imageOffset);
+    const auto imageExtentInBlocks = dstImageTexelBlockInfo.convertTexelsToBlocks(imageExtent);
```

Comment:
you can assert that the offset is a multiple of block size and that extent either is a multiple or extent+offset equal subresourceSize

Reply:
same stuff with granularities


unless you already do it in the `getMemoryNeededForRemainingRegions` method